Add methods to access the wrapped iterator of `Cloned`

tbu- opened the pull request:
I wanted to call `.as_slice()` on a `slice::Iter`, but the iterator was wrapped in a `.cloned()` call.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9856878503/job/27214813656) [(plain)](https://github.com/rust-lang/rust/commit/163a527809e4cad074d7f113975b1e5c1f8fe9c4/checks/27214813656/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#12 [ 5/11] RUN npm install es-check@6.1.1 eslint@8.6.0 -g
#12 5.046 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 5.107 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 5.113 npm WARN deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
#12 5.161 npm WARN deprecated @humanwhocodes/config-array@0.9.5: Use @eslint/config-array instead
#12 5.206 npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
#12 5.206 npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
#12 5.217 npm WARN deprecated @humanwhocodes/object-schema@1.2.1: Use @eslint/object-schema instead
#12 6.246 added 205 packages in 6s
#12 6.246 
#12 6.246 20 packages are looking for funding
#12 6.246   run `npm fund` for details
---
#16 3.479 Building wheels for collected packages: reuse
#16 3.480   Building wheel for reuse (pyproject.toml): started
#16 3.802   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.803   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#16 3.803   Stored in directory: /tmp/pip-ephem-wheel-cache-qiwcmllf/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#16 3.806 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#16 3.828   Attempting uninstall: setuptools
#16 3.828     Found existing installation: setuptools 59.6.0
#16 3.830     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9857079586/job/27215421519) [(plain)](https://github.com/rust-lang/rust/commit/2a1dace2a6d234ea282c298e21565e0d25f70358/checks/27215421519/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#12 [ 5/11] RUN npm install es-check@6.1.1 eslint@8.6.0 -g
#12 4.598 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 4.655 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 4.656 npm WARN deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
#12 4.686 npm WARN deprecated @humanwhocodes/config-array@0.9.5: Use @eslint/config-array instead
#12 4.718 npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
#12 4.733 npm WARN deprecated @humanwhocodes/object-schema@1.2.1: Use @eslint/object-schema instead
#12 5.659 
#12 5.659 added 205 packages in 5s
#12 5.660 
#12 5.660 20 packages are looking for funding
---
#16 3.571 Building wheels for collected packages: reuse
#16 3.572   Building wheel for reuse (pyproject.toml): started
#16 3.893   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.894   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#16 3.895   Stored in directory: /tmp/pip-ephem-wheel-cache-_xh_fgu7/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#16 3.897 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#16 3.919   Attempting uninstall: setuptools
#16 3.919     Found existing installation: setuptools 59.6.0
#16 3.920     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
    Checking core v0.0.0 (/checkout/library/core)
   Compiling libc v0.2.155
   Compiling memchr v2.5.0
   Compiling std v0.0.0 (/checkout/library/std)
error[E0545]: `issue` must be a non-zero numeric string or "none"
   |
   |
27 |     #[unstable(feature = "cloned_inner", issue = "0")]
   |                                                  |
   |                                                  |
   |                                                  `issue` must not be "0", use "none" instead

error[E0545]: `issue` must be a non-zero numeric string or "none"
   |
   |
32 |     #[unstable(feature = "cloned_inner", issue = "0")]
   |                                                  |
   |                                                  |
   |                                                  `issue` must not be "0", use "none" instead

error[E0545]: `issue` must be a non-zero numeric string or "none"
   |
   |
37 |     #[unstable(feature = "cloned_inner", issue = "0")]
   |                                                  |
   |                                                  |
   |                                                  `issue` must not be "0", use "none" instead
error: method has missing stability attribute
  --> library/core/src/iter/adapters/cloned.rs:28:5
   |
28 | /     pub fn get_ref(&self) -> &I {
```

</details>

tbu- committed:
Add methods to access the wrapped iterator of `Cloned`

I wanted to call `.as_slice()` on a `slice::Iter`, but the iterator was
wrapped in a `.cloned()` call.

the8472 commented:
Imo dconstructing adapters does not set a good precedent because ultimately it's not possible to do this for all adapters or would constrain future implementation changes. And since it wouldn't be available for many adapters it wouldn't carry well across refactorings for users.

So it'd be better to encourage the use of `by_ref()` instead which lets you recover the inner iterator once you drop the ref.

See https://github.com/rust-lang/libs-team/issues/128 for a general discussion on the topic

tbu- commented:
> Imo dconstructing adapters does not set a good precedent because ultimately it's not possible to do this for all adapters or would constrain future implementation changes.

This seems more like a slippery slope argument. I think on `Cloned`, this does not constrain future implementation changes because there's just one canonical implementation for `Cloned`.

> So it'd be better to encourage the use of `by_ref()` instead which lets you recover the inner iterator once you drop the ref.

The iterator is part of my data structure, so this would need self-referential types to work.

the8472 commented:
> The iterator is part of my data structure

It doesn't have to be. Cloned has no state, so you can keep the slice iterator and build a `by_ref().cloned()` on the fly when you need an iterator that emits owned items.

tbu- commented:
Indeed, it doesn't have to be. But this means that `Cloned` (or other iterator implementations) are essentially unusable for putting them into data structures if you also need to call methods on the underlying iterator. You can work around that by reimplementing what `Cloned` does.

But maybe the standard library also wants the iterator adapters to be embeddable into data structures (and not just reimplemented everywhere).

the8472 commented:
> But maybe the standard library also wants the iterator adapters to be embeddable into data structures

And there's the slope :wink: . If we add it there'd be questions why it's on this adapter and not some other one. Making it a general feature which is incompatible with some adapters. E.g. `StepBy` cannot expose its internals since they get mangled.

>  this does not constrain future implementation changes because there's just one canonical implementation for Cloned.

That's not correct. With associated type specialization structs can substitute their default fields with optimized representations for specific types, which would be incompatible with returning a reference to the inner iterator.

And as the StepBy specialization shows that with some limitations we don't even need to wait for the associated type kind to do some creative reinterpretation of inner structs. And since such transformations can be slightly lossy it would also make  `into_inner()` impossible.

So it very much does bind the future.

----

To clarify, I do get that this is quite a useful feature. The standard library itself even unpacks iterator abstractions in various places. It's just that the usefulness-to-users conflicts with useful-to-the-implementation properties to have exclusive access to private internals.

So if possible it's preferable that people use `by_ref()` instead to not tie our hands. If there's an important use-case that's impossible to do with `by_ref()` then we can weigh it against giving up implementation flexibility.

tbu- commented:
> And as the StepBy specialization shows that with some limitations we don't even need to wait for the associated type kind to do some creative reinterpretation of inner structs. And since such transformations can be slightly lossy it would also make `into_inner()` impossible.
> 
> So it very much does bind the future.

I see. So the question is really: Can we imagine an optimized implementation for `Cloned` that a) does not contain the original iterator (for the reference methods) or b) that cannot reconstruct the original iterator (for `into_inner`). I see why this is a hard question to ask, especially for *all future optimizations*.

> So if possible it's preferable that people use `by_ref()` instead to not tie our hands. If there's an important use-case that's impossible to do with `by_ref()` then we can weigh it against giving up implementation flexibility.

That's not possible in my case. I can construct the iterator on the fly, but then I'd lose out on the optimization benefits. So the correct solution might just be to reimplement all the iterator abstractions on my own so that I can also access the inner iterator.

the8472 commented:
> I see. So the question is really: Can we imagine an optimized implementation for Cloned that a) does not contain the original iterator (for the reference methods) or b) that cannot reconstruct the original iterator (for into_inner). I see why this is a hard question to ask, especially for all future optimizations.

That is quite hard indeed. Not only does it depend on future optimizations but also future iterator types that might require new optimizations.

A simple "existence proof" for a) involves lending iterators. When you have an unfortunate API that takes a lending iterator but the underlying source is owning then we may be able to excise the step that takes an owned item and turns it into a short-lived reference which the consumer then clones and just return the original item. By unpacking the iterator two levels deep this would no longer contain the field that temporarily holds the owned value that's being lent.

b) might involve somthing reference-counted where if `Cloned` knows it is the last, exclusive owner it can just steal the values instead of cloning them. Accessing original source could then lead to surprising results, depending on which APIs the  source exposes. We have no such thing in std today afaik.

> but then I'd lose out on the optimization benefits

Which ones are relevant for you? `Cloned` is somewhat less optimized than `Copied` anyway. I think it's based on the assumption that clone impls will do things that are more likely to inhibit optimizations.  `TrustedLen` does propagated through `by_ref()`.

tbu- commented:
For future optimization potential, we don't want to guarantee that `Cloned` contains its inner iterator or can even reconstruct it. Since my use case has a work-around (manually implement `Cloned` in my code), the decision is that we don't want this PR.
